### PR TITLE
Add ImageDecoder operator

### PR DESCRIPTION
**Pull request by jantonguirao**
### Description

- Adds a new ImageDecoder operator, to be used in preprocessing models.
- Tests use generated images and opencv as a reference.

### Motivation and Context

Currently, image decoding happens outside of ONNX, bring problems of portability and deployment.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 numpy
 protobuf >= 3.20.2
+opencv-python
```
Comment (xadupre) [inline requirements.txt:3]: We should make it optional.

File: requirements-release.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 numpy==1.24.3
 protobuf==4.21.12
+opencv-python
```
Comment (xadupre) [inline requirements-release.txt:3]: We should make it optional.

File: onnx/defs/image/defs.cc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+static const char* ImageDecoder_ver20_doc =
+    R"DOC(Loads and decodes and image from a file. If it can't decode for any reason (e.g. corrupted encoded stream, invalid format, it will return an empty matrix).
+
+The following image formats are supported: BMP, JPEG, JPEG2000, TIFF, PNG, WebP, Portable image format (PBM, PGM, PPM, PXM, PNM).
+
+Decoded images follow a channel-last layout: (Height, Width, Channels).
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply onnx/defs/image/defs.cc:20]: I believe we discussed that earlier, and the conclusion was not to add a domain unless we have a significant number of operators that would justify it.

File: requirements.txt (hunk 1 above)
Comment (jantonguirao) [reply requirements.txt]: Done. I used [PEP-508](https://peps.python.org/pep-0508/#extras). Would that be enough?

File: requirements-release.txt (hunk 2 above)
Comment (jantonguirao) [reply requirements-release.txt]: Done

File: requirements.txt (hunk 1 above)
Comment (jantonguirao) [reply requirements.txt]: Will try removing the dependency from here completely.

File: onnx/defs/image/defs.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnx/defs/image/defs.cc:20]: Maybe a conclusion was reached somewhere else but to me this op feels very different than the other ones we use to create neural networks. It seems it would make sense mentally that a backend doesn't need to implement a decoding op for it to support core ONNX?

PyTorch and tensorflow both have image decoding operations scoped in a different namespace.

[the remark this answers is not in the document]
Comment (jantonguirao) [reply onnx/defs/image/defs.cc:20]: I do agree with your points above. This operator seems special enough to justify a new "image" domain. I'll consider it and bring it for discussion with the preprocessing working group. Thanks.

[the remark this answers is not in the document]
Comment (xadupre) [reply onnx/defs/image/defs.cc:20]: It is one operator now. What happens when the number is big enough to justify a new domain? Or maybe we could see that as an IO operator. We may have other similar operators to convert compress data into tensors, data being images, sound, video, ...

[the remark this answers is not in the document]
Comment (gramalingam) [reply onnx/defs/image/defs.cc:20]: I think one  important point is Justin's comment "a backend doesn't need to implement a decoding op for it to support core ONNX". More generally, what is the expectation/goal/intention of using other namespaces? Is there an implication that this op is not as important as the ones in the standard domain, and that core onnx backends may not implement it?

[the remark this answers is not in the document]
Comment (gramalingam) [reply onnx/defs/image/defs.cc:20]: What would be nice is some set of principles/guidelines for making the decision about which domain to place new ops in, something we haven't had to deal with so far (except for onnx vs onnxml, where there is some principle/guideline).

[the remark this answers is not in the document]
Comment (gramalingam) [reply onnx/defs/image/defs.cc:20]: Should it be generic like preprocessing or prepostprocessing? Or, specific to image?

File: onnx/defs/image/defs.cc [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        .SetDoc(ImageDecoder_ver20_doc)
+        .Attr(
+            "pixel_format",
+            "Pixel format. Can be one of \"RGB\", \"BGR\", or \"Grayscale\".",
+            AttributeProto::STRING,
+            std::string("RGB"))
+        .Attr(
+            "dtype",
+            "(Optional). Specifies data type for the elements of the output tensor T2. "
+            "Meant for future extensions, currently limited to uint8 only. "
```
Comment (gramalingam) [inline onnx/defs/image/defs.cc:33]: Why add this now? It seems better that this attribute and the TypeConstraint for T2 be consistent with each other. So, either support other types throughout the spec or omit the attribute?

File: onnx/defs/image/defs.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+          }
+          if (ctx.getAttribute("dtype") != nullptr) {
+            propagateElemTypeFromDtypeToOutput(ctx, ctx.getAttribute("dtype"), 0);
+          } else {
+            propagateElemTypeFromDtypeToOutput(ctx, TensorProto::UINT8, 0);
+          }
+          auto output_type = ctx.getOutputType(0);
+          auto* sh = output_type->mutable_tensor_type()->mutable_shape();
+          sh->clear_dim();
+          sh->add_dim()->set_dim_param(std::string("H"));
```
Comment (gramalingam) [inline onnx/defs/image/defs.cc:70]: It is better to leave the dim-param unset. The dim-params like "H" have a global scope. This would make it look like two different invocations of this op return the same dimensions for their outputs.

File: onnx/defs/image/defs.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (xadupre) [reply onnx/defs/image/defs.cc:20]: `ai.onnx.io` for anything related to images, sound, videos, ... ?

[the remark this answers is not in the document]
Comment (jantonguirao) [reply onnx/defs/image/defs.cc:20]: It makes more sense to me to use `image` (and include image processing operations in the future) or `image_codec` (only image decoding/encoding) as domain. I can imagine a backend that wants to implement anything relate to image processing (including decoding) but has no means to implement audio/video operators. 

I agree that we need to decide and document our approach and implications of using domains.

File: onnx/defs/image/defs.cc (hunk 4 above)
Comment (jantonguirao) [reply onnx/defs/image/defs.cc]: Removed it.

**Comment by jantonguirao** [comment]
As agreed during the Operators SIG meeting, we will not use a separate domain for this operator, at least for the time being.

File: onnx/defs/image/defs.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnx/defs/image/defs.cc:39]: The decision for this PR is to leave it in the standard domain. We can consider adding ops to new domains once we have  (i) Infrastructure support (existing version-converter and function-op-definition mechanism both support the standard domain specially, but do not handle other domains), and (ii) There is a concrete proposal (about 1 or more new domains to be added, as well as the conventions to be used to decide when to put which op in which domain).

**Review by gramalingam — APPROVED** [review]

File: requirements-dev.txt [hunk 6]
```diff
@@ -1,5 +1,6 @@
 numpy
 protobuf
+opencv-python
```
Comment (xadupre) [inline requirements-dev.txt:3]: It should not be needed since it is also added at the last line.

Comment (jantonguirao) [reply requirements-dev.txt]: True. Fixed

**Review by xadupre — APPROVED** [review]

File: requirements-dev.txt [hunk 7]
```diff
@@ -14,3 +14,5 @@ lintrunner-adapters>=0.8
 # Dependencies for the reference implementation.
 google-re2
```
Comment (adityagoel4512) [inline requirements-dev.txt:15]: Can you delete this line now that you have added the line `-r requirements-reference.txt`

Comment (jantonguirao) [reply requirements-dev.txt]: Done

**Comment by jantonguirao** [comment]
> Hi @jantonguirao , thank you for adding this op. Now I am preparing ONNX 1.15.0 release and integrating the release with ORT, I wonder if you have a plan to implement the runtime kernel for this new op in ORT? If so, I will let you know once ORT has ONNX 1.15.0 integrated. Thank you in advance.

Hi @liqunfu. I was hoping that the ORT team will handle the implementation. It should be simple using a library such as OpenCV.

**Comment by liqunfu** [comment]
> > Hi @jantonguirao , thank you for adding this op. Now I am preparing ONNX 1.15.0 release and integrating the release with ORT, I wonder if you have a plan to implement the runtime kernel for this new op in ORT? If so, I will let you know once ORT has ONNX 1.15.0 integrated. Thank you in advance.
> 
> Hi @liqunfu. I was hoping that the ORT team will handle the implementation. It should be simple using a library such as OpenCV.

Thank you @jantonguirao , I will add this to ORT release work item and have ORT team work on it.
